### ISSUE TITLE
Are obvious ambigous functions declarations only ambigous at call time or before

### ISSUE THREAD

**Issue by apinski-cavium** [p4lang/p4c#3192]
Take:
```
bit<2> add1(bit<2> t)
{
  return t + 1;
}
bit<2> add1(bit<2> t)
{
  return t + 2;
}
bit<2> f(bit<2> t)
{
  return add1(t);
}
```
The two declarations of add1 are obviously ambigous but the reference compiler rejects it only at the call site. The specification talks a little bit about function overloading but really only that generics can't add an implicit casts.
This area definitely needs some expansion just to make it understand that if there was no call to add1, the always ambigous call will never be error'ed out (the reference compiler should warn about this case).

**Comment by mbudiu-vmw** [comment]
Yes, that is right, we only check if a function or method is ambiguous when we try to resolve it.
If the resolution rules produce multiple possible candidates we give an error message.
This case is simpler, but in the presence of generic functions it's harder to answer the question of whether two functions could be ambiguous. That's why we only check when the functions are invoked.
If you have a proposal of an algorithm to do this eagerly I am open to implementing it.

**Comment by mbudiu-vmw** [comment]
I am pretty sure there was another issues about this, but I am not sure I can easily find it.
